finality-aleph/src/testing/network.rs

```diff
@@ -0,0 +1,315 @@
+            expected_addresses.extend(authority.addresses());
+        }
+
+        assert_eq!(reserved_addresses, expected_addresses);
+    }
+
+    async fn check_sends_authentication(&mut self, authentication: Authentication) {
+        let mut sent_auth = HashMap::new();
+        while sent_auth.len() < NODES_N - 1 {
+            if let Some((
+                NetworkData::Meta(DiscoveryMessage::Authentication(auth_data)),
+                peer_id,
+                protocol,
+            )) = timeout(DEFAULT_TIMEOUT, self.network.send_message.next())
+                .await
+                .expect("Should send authentication")
+            {
+                assert_eq!(protocol, Protocol::Generic.name());
+                sent_auth.insert(peer_id, auth_data);
+            }
+        }
+
+        let mut expected_auth = HashMap::new();
+        for authority in self.authorities.iter().skip(1) {
+            expected_auth.insert(authority.peer_id(), authentication.clone());
+        }
+
+        assert_eq!(sent_auth, expected_auth);
+    }
+
+    async fn connect_session_authorities(&mut self, session_id: u32) {
+        for (index, authority) in self.authorities.clone().into_iter().enumerate().skip(1) {
+            let handler = self.get_session_handler(index, session_id).await;
+
+            self.connect_identity_to_network(authority.peer_id(), Protocol::Generic);
+            self.connect_identity_to_network(authority.peer_id(), Protocol::Validator);
+
+            self.network.emit_event(Event::NotificationsReceived {
+                remote: authority.peer_id().into(),
+                messages: vec![(
+                    Protocol::Generic.name(),
+                    NetworkData::<MockData>::Meta(DiscoveryMessage::AuthenticationBroadcast(
+                        handler.authentication().unwrap(),
+                    ))
+                    .encode()
+                    .into(),
+                )],
+            });
+        }
+    }
+
+    async fn start_session(&mut self, session_id: u32) -> SessionNetwork<MockData> {
+        let data_network = self.start_validator_session(0, session_id).await;
+        self.connect_session_authorities(session_id).await;
+        self.check_sends_add_reserved_node().await;
+        self.check_sends_authentication(
+            self.get_session_handler(0, session_id)
+                .await
+                .authentication()
+                .unwrap(),
+        )
+        .await;
+
+        data_network
+    }
+
+    fn emit_notifications_received(
+        &mut self,
+        node_id: usize,
+        messages: Vec<NetworkData<MockData>>,
+    ) {
+        self.network.emit_event(Event::NotificationsReceived {
+            remote: self.authorities[node_id].peer_id().into(),
+            messages: messages
+                .iter()
+                .map(|m| (Protocol::Validator.name(), m.encode().into()))
+                .collect(),
+        });
+    }
+
+    async fn cleanup(self) {
```

Comment:
It would be a bit easier to read and understand how to use it if you would introduce some access/visibility modifiers for these methods, instead of making everything public by default.

Reply:
I use all of those methods in tests. They were kind of designed to work by themselves.